TF-M based PKCS#11 library implementation

Sherryzhang2 opened the pull request:
TF-M based PKCS#11 library implementation

Description:
* Implement the PKCS#11 APIs based on TF-M.
* The iot_pkcs11_pal layer is removed. The objects are stored in TF-M via PSA API. The platform dependency of storage is implemented in TF-M other than Amazon FreeRTOS.
* Interfaces of converting MBEDTLS format input into PSA format input are added.
* Add PKCS#11 library test project.

TODO: 
Use the Amazon-FreeRTOS logging task for printing logs instead of calling Musca-B1 platform print APIs.

CharleyChu commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,2337 @@
+							}
+							else if( xPalHandle == eAwsDevicePublicKey )
+							{
+								ulKeyHandle = P11KeyConfig.uxDevicePrivateKey;
```
should be 
ulKeyHandle = P11KeyConfig.uxDevicePublicKey;

Sherryzhang2 replied:
yes, I will fix it in the next commit.

CharleyChu commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,2337 @@
+							psa_key_handle_t ulKeyHandle = 0;
+							if( xPalHandle == eAwsDevicePrivateKey )
+							{
+								ulKeyHandle = P11KeyConfig.uxDevicePublicKey;
```
should be 
ulKeyHandle = P11KeyConfig.uxDevicePrivateKey;

Sherryzhang2 replied:
yes, I will fix it in the next commit.

dcgaws commented on demos/dev_mode_key_provisioning/src/aws_dev_mode_key_provisioning.c:
```diff
@@ -921,8 +918,8 @@ static void prvWriteHexBytesToConsole( CK_SESSION_HANDLE xSession,
             *pcNextChar = '\0';
-            vLoggingPrint( pcByteRow );
-            vLoggingPrint( "\r\n" );
+            print_log( pcByteRow );
```
Looks like this is a gap in the #defines we provide in the reference. We should offer one for printing naked log lines, rather than requiring you to make these modifications.

Sherryzhang2 replied:
In the commit message, the print log issue is mentioned in the TODO list. The plan is to use the longing task in libraries/c_sdk, thus the gap can be dismissed. I will handle the issue in the next commit.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+				break;
+
+			case ( CKA_MODULUS ):
+				lMbedReturn |= mbedtls_rsa_import_raw( pxRsaContext,
```
Why OR this in?

Sherryzhang2 replied:
The intention is to gather the errors generated in checking the necessary components of the key. At the end of the function, the CKR_ATTRIBUTE_VALUE_INVALID error is returned if any error occurred during the check.

dcgaws replied:
It would be more efficient to abort the loop after the first error, then.

Sherryzhang2 replied:
Make sense. I will update it in the next patch.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+						  CK_ULONG ulCount,
+						  CK_OBJECT_HANDLE_PTR pxObject )
+{
+#define MAX_PUBLIC_KEY_SIZE	3000
```
Is it the intent of this port to support 2048-bit RSA and ECDSA P256?

Sherryzhang2 replied:
yes, I will add the comment in the code in the next commit.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+	return xResult;
+}
+
+CK_RV prvMbedTLS_Initialize( void )
```
Does this code run on the main MCU side, or on the secure MCU? 

Would I be correct in assuming that the secure side is using mbed crypto?

Sherryzhang2 replied:
TF-M runs on a single ARMv8 core. Yes, it is correct that the secure side is using mbed crypto which is called by the PSA crypto service and is initialized by TF-M. The code here is part of the PKCS11 library which runs on the Non-Secure side. The Non-Secure side calls the mbedtls functions for parsing a key or certificate.

michaelthomasj replied:
This shim layer makes calls to mbedtls_xxx crypto functions (which are in mbedCrypto) to perform crypto operations. According to your comment above, the mbedCrypto code is in secure area and exposed via PSA Crypto Service. 
So how will the PKCS11 layer which is in non-secure area call the mbedtls_xxx crypto functions which are in secure area? Do you intend to expose all those functions also via the PSA Crypto service?

david-wang-2015 replied:
Hi Michael,
The PKCS#11 shim layer running at the non-secure side calls the PSA Storage and Crypto APIs to the secure side which runs TF-M.
The implementation of the PSA Crypto API in TF-M is currently mbedcrypto. But it doesn't have to be mbedcrypto as the secure side backend or even TF-M itself. This shim layer can support any PSA API compliant implementation which may use SW or HW crypto engine.
Hope this helps to clarify the relationship between PSA Crypto API (service) and mbedcrypto.

michaelthomasj replied:
Hi David, 
Maybe I dont undertand the file partitioning, but if you look at "prvCreateEcPrivateKey()" there is a function in there called "mbedtls_mpi_read_binary()". Where does mbedtls_mpi_read_binary() reside? I know its in bignum.c, and when I used mbedCrypto it was part of that codebase. So the assumption I made is that that function is in the secure area. If so, then how will this shim layer which is in non-secure area call that function directly instead of through a service?

david-wang-2015 replied:
Hi Michael,
You're right. The "mbedtls_mpi_read_binary()" is for key format conversion. It's indeed in the non-secure side as the helper function.
For our shim layer, it's basically:
* Maps PKCS#11 object storage (just certificate and key) to PSA Storage API (the backend of PSA Storage API is TF-M storage service in the example)
* Maps PKCS#11 Crypto API (just the ones used by AWS TLS library, mainly sign function) to PSA Crypto API (the backend of PSA Crypto API is TF-M crypto service which uses its own mbedcrypto)

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+		/*
+		 * Clean-up find objects state.
+		 */
+		vPortFree( pxSession->pxF
```
If there are multiple tasks performing hashing in parallel on the main MCU (i.e. via PKCS11), does the PSA side handle that?

Sherryzhang2 replied:
Yes, the PSA supports multiple Non-Secure side tasks performing hash operations in parallel. But currently, TF-M does not support the multi-context. To solve this issue, TF-M only supports Non-Secure task call PSA service in sequence. A mutex lock is used to block other Non-Secure tasks calling PSA service until no other task is executing on the secure side. TF-M will support multi-context later.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+		/*
+		 * Clean-up find objects state.
+		 */
+		vPortFree( pxSession->pxF
```
Just to confirm, this port appears to be supporting RSA and ECDSA runtime operations, but only ECDSA for onboard keygen. Is that intended?

Sherryzhang2 replied:
PSA also supports the other keygens. This port intends to keep consistent with the mbedtls based PKCS#11 implementation in Amazon FreeRTOS which only supports the ECDSA keygen on this version. So only the ECDSA keygen is supported in this port.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+		/*
+		 * Clean-up find objects state.
+		 */
+		vPortFree( pxSession->pxF
```
Out of curiosity, does the PSA side implement a TRNG in hardware?

Sherryzhang2 replied:
The platform port of Musca-B1 in this version of TFM does not use the hardware to get the TRNG. However, the latest version(TF-Mv1.0-RC3) TF-M supports CryptoCell-312 which is used to get the random number generated from the hardware. In general, we leave it up to each platform to implement the TRNG.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa_object_management.c:
```diff
@@ -0,0 +1,600 @@
+					pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS ) == 0 ) )
+		{
+			/*
+			 * Currently TF-M does not support persistent key APIs, so volatile
```
Does this comment apply to the whole PKCS11 port herein, i.e., that persistent private keys aren't supported?

Sherryzhang2 replied:
Yes, the comment still applies. The persistent private keys have not been supported in TF-M yet. It is still under development now.

dcgaws commented on vendors/arm/boards/musca-b1/aws_tests/application_code/arm_code/iot_test_pkcs11_psa.c:
```diff
@@ -0,0 +1,53 @@
+int mbedtls_hardware_poll( void *data,
+						   unsigned char *output, size_t len, size_t *olen )
+{
+	/* FIX ME
```
Is there a plan to fix this in the current milestone?

Sherryzhang2 replied:
This function is only used in the PKCS#11 test case. In the PKCS#11 test, it calls the mbedtls steps to generate the random number, so this function is needed. But in the PKCS#11 library, we call the C_GenerateRandom to generate a random number and do not need to call this function. I will remote the FIX ME  comment and add the explanation here.

dcgaws commented on vendors/arm/boards/musca-b1/aws_tests/config_files/FreeRTOSConfig.h:
```diff
@@ -0,0 +1,286 @@
+/* The secure stack is managed by TFM. So define the macro to 0. */
+#define configMINIMAL_SECURE_STACK_SIZE                 0
+
+#define configUSE_DAEMON_TASK_STARTUP_HOOK         0
```
Would configUSE_MPU be applicable here?

Sherryzhang2 replied:
Actually no macro configUSE_MPU is defined. The macro that controls whether MPU is used is configENABLE_MPU which is set at line 278 of this file.

dcgaws commented on vendors/arm/boards/musca-b1/aws_tests/config_files/aws_test_runner_config.h:
```diff
@@ -0,0 +1,55 @@
+#define testrunnerFULL_OTA_CBOR_ENABLED            testrunnerUNSUPPORTED
+#define testrunnerFULL_OTA_AGENT_ENABLED           testrunnerUNSUPPORTED
+#define testrunnerFULL_OTA_PAL_ENABLED             testrunnerUNSUPPORTED
+#define testrunnerFULL_MQTT_ALPN_ENABLED           testrunnerUNSUPPORTED
```
It would be helpful to run at least one of the test groups that performs mutual authentication with AWS IoT. Realistically, that can find crypto port bugs that aren't yet caught by the PKCS11 tests.

Sherryzhang2 replied:
Yes. Actually we also plan to port the TCP/IP stack of Amazon FreeRTOS to arm platform Musca-B1 to connect to Amazon echo server. So that we can test the mutual authentication.

dcgaws commented on vendors/arm/boards/musca-b1/aws_tests/config_files/iot_test_pkcs11_config.h:
```diff
@@ -0,0 +1,141 @@
+ * Each task consumes both stack and heap space, which may cause memory allocation
+ * failures if too many tasks are created.
+ */
+#define pkcs11testMULTI_THREAD_TASK_COUNT     ( 2 )   /* FIXME. */
```
Do the "FIXME" comments still apply?

Sherryzhang2 replied:
Not anymore. I will remove it in the next commit.

dcgaws reviewed (changes requested):
I left a handful of inline comments. Will you please take a look? In the meantime, thank you for initiating this work.

CharleyChu commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa_object_management.c:
```diff
@@ -0,0 +1,600 @@
+	psa_ps_status_t uxStatus;
+	psa_ecc_curve_t curve_id;
+	const mbedtls_ecp_keypair *ec;
+	unsigned char *pcPrivateKeyDerStart = NULL;
```
define buffer for raw private key (see comment at line 198
        unsigned char pcPrivateKeyRaw[MBEDTLS_ECP_MAX_BYTES];
        size_t xPrivateKeySizeRaw = 0;

Sherryzhang2 replied:
See the reply at line 198.

CharleyChu commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa_object_management.c:
```diff
@@ -0,0 +1,600 @@
+						curve_id = mbedtls_ecp_curve_info_from_grp_id( ec->grp.id )->tls_id;
+						uxKeyType = PSA_KEY_TYPE_ECC_KEYPAIR(curve_id);
+						uxAlgorithm = PSA_ALG_ECDSA( PSA_ALG_SHA_256 );
+						if ( 0 != get_ECPrivateKey_sequence_privatekey_sec1_der(
```
With DER format, a leading zero will be added if the MSB of key is "1",  the key size will be one byte more than the algorithm key size. In this case, the key will be rejected the key length doesn't match. Contacted mbed-crypto developer,  I was told only raw (binary) format of private key is supported for psa_import_key().

Here is my change for that
                                                ec = mbedtls_pk_ec(*pvContext);;
                                                curve_id = mbedtls_ecp_curve_info_from_grp_id( ec->grp.id )->tls_id;
                                                uxKeyType = PSA_KEY_TYPE_ECC_KEYPAIR(curve_id);
                                                uxAlgorithm = PSA_ALG_ECDSA( PSA_ALG_SHA_256 );
                                                xPrivateKeySizeRaw = ( ec->grp.nbits + 7 ) / 8;
                                                if( mbedtls_mpi_write_binary( &ec->d, pcPrivateKeyRaw, xPrivateKeySizeRaw ) != 0 )
                                                {
                                                        uxStatus = PSA_ERROR_GENERIC_ERROR;
                                                }
                                                pucKeyData = pcPrivateKeyRaw;
                                                ulKeyDataSize = xPrivateKeySizeRaw;

Sherryzhang2 replied:
First of all, thanks for your propose here.
I am sorry that the name of the function get_ECPrivateKey_sequence_privatekey_sec1_der is a lit confusing. Let me explain the solution I used. This function is to extract the privateKey and its length from the key in DER format. When encoding the key into DER binary format, the privateKey field is encoded by calling the  mbedtls_asn1_write_mpi function in which the length of the privateKey field is calculated(as you described, ( ( mbedtls_mpi_bitlen( ec->d ) + 7 ) >> 3 ) ) and the data length of stored in the address next to the privateKey encoded data. The data length, as well as the privateKey field, is encoded as the MBEDTLS_ASN1_OCTET_STRING tag. The function get_ECPrivateKey_sequence_privatekey_sec1_der is to extract the privateKey field binary data and its length. It looks for the address where the privateKey(as well as its length) is stored and then get the length of the tag MBEDTLS_ASN1_OCTET_STRING which is exactly the privateKey length. Thus we get the start address and the length of the privateKey binary format which is exactly what we need.
So your proposal is very similar to mine except for the data length input.
The input data length should be the binary format privateKey's data length. The psa_import_key uses the data start address and its length to check whether the input data is valid. So I think it isn't reasonable the take ( ec->grp.nbits + 7 ) / 8 as the input data length. Input data length should be the length of what your really input anyway.

CharleyChu replied:
With your original code, we see error while running AFR IDT test. the private key is rejected as key size mis-match (a leading zero in DER format, resulting in the key size is 33 bytes, instead of 32 bytes). I spoken to mbedcrypto developer, they say PSA SST only support raw format now. So I made the change to convert the key in raw format.

Sherryzhang2 replied:
yes, you are right. The raw binary format of privateKey should be taken as the input parameter. I have updated the code to fix it.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+						   CK_ULONG ulCount,
+						   CK_OBJECT_HANDLE_PTR pxObject )
+{
+	/* TODO: How long is a typical RSA key anyhow?  */
```
Won't this be 256 bytes (i.e. for a binary encoded RSA private)? Or maybe a little more than that if there's a header? 3k is a huge amount of RAM.

Sherryzhang2 replied:
For RSA-2048, the modulus n is 2048bits(256 bytes). But here it should be the length of the key in DER format. The content of the RSA private key is as follows:
RSAPrivateKey ::= SEQUENCE {
  version           Version,
  modulus           INTEGER,  -- n   (256 bytes for RSA-2048)
  publicExponent    INTEGER,  -- e  (3 or 4 bytes)
  privateExponent   INTEGER,  -- d  (about 256 bytes)
  prime1            INTEGER,  -- p  (128 bytes)
  prime2            INTEGER,  -- q  (128 bytes)
  exponent1         INTEGER,  -- d mod (p-1)  (128 bytes)
  exponent2         INTEGER,  -- d mod (q-1)  (128 bytes)
  coefficient       INTEGER,  -- (inverse of q) mod p  (128 bytes)
  otherPrimeInfos   OtherPrimeInfos OPTIONAL
}
Besides, when encode the key, some ASN tags are also introduced.
For RSA-2048, the encoded length should be around 1232  bytes.

For EC key the max der format private key length is less than 1232 bytes.

So I will change be value to 1300 bytes.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+
+				if( xBool != CK_TRUE )
+				{
+					PKCS11_PRINT( ( "Only EC Public Keys with verify permissions supported. \r\n" ) );
```
Should the presence of this attribute in the template be mandatory then? Ditto for CKA_SIGN in the case of a private key, for example.

Sherryzhang2 replied:
Yes, just keep consistent with the MbedTLS based implementation.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+						  CK_ULONG ulCount,
+						  CK_OBJECT_HANDLE_PTR pxObject )
+{
+#define MAX_PUBLIC_KEY_SIZE	3000	/* To cover 2048-bit RSA and ECDSA P256 type public key. */
```
Similar to above - this gives customers the impression that this port incurs a huge RAM overhead.

Sherryzhang2 replied:
For RSA-2048 public key, the content is:

RSAPublicKey ::= SEQUENCE {
    modulus           INTEGER,  -- n   (256 bytes)
    publicExponent    INTEGER   -- e (3 or 4 bytes)
}

Besides, when encoding the key, some ASN tags are also introduced.
For RSA-2048, the encoded length should be around 310 bytes.

For EC key the max der format private key length is less than 310 bytes.

So I will change be value to 330 bytes.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+	if( xResult == CKR_OK )
+	{
+		memset( &xP11Context, 0, sizeof( xP11Context ) );
+		xP11Context.xObjectList.xMutex = xSemaphoreCreateMutex();
```
This call can fail.

Sherryzhang2 replied:
Yes, I will add the handle of the failure case.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+	{
+		{ CKM_RSA_PKCS,		{ 2048, 2048, CKF_SIGN			  } },
+		{ CKM_RSA_X_509,	   { 2048, 2048, CKF_VERIFY			} },
+		#if ( pkcs11configSUPPRESS_ECDSA_MECHANISM != 1 )
```
I suspect you don't want this in your port.

Sherryzhang2 replied:
This port only supports the unsuppressed key mechanism. This macro is always not defined.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+	if( xResult == CKR_OK )
+	{
+		pxFindObjectLabel = pvPortMalloc( pxTemplate->ulValueLen + 1 ); /* Add 1 to guarantee null termination for PAL. */
+		pxSession->pxFindObjectLabel = pxFindObjectLabel;
```
It would be safer to move this line into the if() block below.

Sherryzhang2 replied:
Make sense. I will refine it in the next patch.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+		}
+	}
+
+	/* TODO: Re-inspect this previous logic. */
```
I think this comment is leftover from the software version? In any case, do you see any problems with the logic it's referring to? It would be good to ensure that we aren't copy & pasting bugs.

Sherryzhang2 replied:
Yes, the comment does not apply now. I will remove it.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+	if( xResult == CKR_OK )
+	{
+		/*
+		 * Clea
```
Isn't pPart a byte array of data to be hashed?

Sherryzhang2 replied:
Yes, it is. The log here is not right. I will fix the log in the next patch.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+	if( xResult == CKR_OK )
+	{
+		/*
+		 * Clea
```
Could just call DigestFinal rather than replicate the clean-up code?

Sherryzhang2 replied:
As the PSA cryptography spec says, psa_hash_finish calculates the hash of the message formed by concatenating the inputs passed to preceding calls to psa_hash_update. If psa_hash_update returns an error status, the operation becomes inactive. So psa_hash_finish will not work properly in this case.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+	if( xResult == CKR_OK )
+	{
+		/*
+		 * Clea
```
This lock might be redundant in the PSA case. Just confirming.

Sherryzhang2 replied:
This lock is used to protect the sign parameters, such as key and sign algorithm is not changed in the signing process. Without the lock, if one task is signing another task change the parameters by calling c_signinit, the signing process will be probably not done successfully. So I think this lock is essential.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -0,0 +1,3476 @@
+	if( xResult == CKR_OK )
+	{
+		/*
+		 * Clea
```
This is redundant to the previous block.

Sherryzhang2 replied:
I am sorry that the code may be a bit confusing. 
The process here is:
Step1: allocate the private key (just get the key handle, the key content has not been created) and set its usage and policy.
Step2: generate the key pair(private and public key).
Step3: export the public key content.
Step4: allocate the public key(just get the key handle) and set its usage and policy.
Step5. import the public key with the key content got in step 3.
Finally, the private key and public key are all stored in the TF-M and their key handles are all returned.

I will add some comments to make the code more friendly.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -46,57 +46,57 @@
-#define pkcs11NO_OPERATION			( ( CK_MECHANISM_TYPE ) 0xFFFFFFFFF )
+#define PKCS11_PRINT( X )            vLoggingPrintf X
+#define PKCS11_WARNING_PRINT( X )    vLoggingPrintf  X
+#define pkcs11NO_OPERATION           ( ( CK_MECHANISM_TYPE ) 0xFFFFFFFFF )
```
Should this be declared with explicit signed-ness?

Sherryzhang2 replied:
Yes, it makes sense. I will update it in the next patch.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -213,135 +213,135 @@ static CK_FUNCTION_LIST prvP11FunctionList =
+    {
+        if( xP11Context.xObjectList.xObjects[ lIndex ].xHandle != CK_INVALID_HANDLE )
+        {
+            *ppcLabel = xP11Context.xObjectList.xObjects[ lIndex ].xLabel;
```
This pointer will be good for the lifetime of the session?

Sherryzhang2 replied:
Yes, the pointer ppcLabel is always the address of a local variable which is a pointer that only used locally. Here only when the xHandle is not CK_INVALID_HANDLE the ppcLable pointer will be assigned.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -354,790 +354,826 @@ void prvFindObjectInListByLabel( uint8_t * pcLabel,
+    if( xResult == CKR_OK )
+    {
+        xResult = prvAddObjectToList( xPalHandle, pxObject, pxLabel->pValue, pxLabel->ulValueLen );
+        /* TODO: If this fails, should the object be wiped back out of flash?  But what if that fails?!?!? */
```
Would it be safer to do the operations in the other order?

Sherryzhang2 replied:
Do you mean that prvAddObjectToList firstly and then PKCS11_PSA_SaveObject? That does not work as the first parameter, xPalHandle, is returned by PKCS11_PSA_SaveObject. I will add the handle of the failure case in the next patch.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -354,790 +354,826 @@ void prvFindObjectInListByLabel( uint8_t * pcLabel,
 void prvGetLabel( CK_ATTRIBUTE_PTR * ppxLabel,
-				  CK_ATTRIBUTE_PTR pxTemplate,
-				  CK_ULONG ulCount )
+                  CK_ATTRIBUTE_PTR pxTemplate,
```
Looks like this and the previous function could be refactored.

Sherryzhang2 replied:
The types of the first input parameter of the two functions are not the same. It would make the parameter unfriendly to users if combine the two functions together. And the code size is not that much. So I prefer to keep it as it is.

dcgaws commented on libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c:
```diff
@@ -354,790 +354,826 @@ void prvFindObjectInListByLabel( uint8_t * pcLabel,
+    *ppxClass = NULL;
+
+    /* Key will be assembled in the mbedTLS key context and then exported to DER for storage. */
+    mbedtls_ecp_keypair * pxKeyPair = ( mbedtls_ecp_keypair * ) pxMbedContext->pk_ctx;
```
Is it desirable to have a hard dependency on mbedTLS here?

Sherryzhang2 replied:
The mbedTLS library provides cryptographic operations such as key format encode, decode, hash certificate parse and so on. The mbedTLS library is used as a help function.  This implementation is not intended to replace the mbedTLS library in the Non-Secure side.

Sherryzhang2 committed:
Add PKCS#11 library test project based on Musca_b1 platform

The PSA implementation is Trusted Firmware M(TF-M). The demo is based on
Arm Musca_b1 platform.

Change-Id: Ica1ad24ff14de9cbfcaa82fb3932ba5a25c2aab1
Signed-off-by: Sherry Zhang <sherry.zhang2@arm.com>

david-wang-2015 commented:
Seems no more comments so far. We're going to consolidate the patches.
On top of the consolidated patch, we will push the persistent key support for your review.
Please let us know if any questions.
Thanks.

Sherryzhang2 committed:
PSA based implementation of PKCS#11 shim layer

This is the Platform Security Architecture(PSA) based implementation
of PKCS#11 shim layer. It follows the "PSA Storage API 1.0 beta 2" and "PSA
Cryptography API 1.0 beta 2". It is platform independent. The PSA firmware
implementation (e.g. Trusted Firmware-M) which follows these APIs can be used
at the secure side.

Change-Id: I6b84953d2b8aaa8471fcd91b168794f9895176d9
Signed-off-by: Sherry Zhang <sherry.zhang2@arm.com>

Sherryzhang2 committed:
Align PSA Cryptographic API V1.0 beta3 and Storage API to V1.0 and enable persistent key

Align to "PSA Storage API 1.0" and "PSA Cryptography API 1.0 beta3".
Change the device keys to persistent key so that users can make use
of the key which is provisioned by others.

Change-Id: I8307b9c0e58cd949618d1a911acced82a273bc4b
Signed-off-by: Sherry Zhang <sherry.zhang2@arm.com>

Sherryzhang2 committed:
Add the ReadMe.txt for the PSA based PKCS#11 shim layer

Change-Id: I746ab5b3a1ab3e147c825e20d6bc029dd8efdcd2
Signed-off-by: Sherry Zhang <sherry.zhang2@arm.com>

david-wang-2015 commented:
Merged this PR.
It passes the PKCS#11 test suites on Arm Musca B1 board.
Further enhancements and bug fixes will be addressed in other PRs.
